### ISSUE TITLE
builder: Proposal: Add `TARGETPLATFORM/BUILDPLATFORM` default ARG to Dockerfile

### ISSUE THREAD

**Issue by tonistiigi** [moby/moby#32487]
I've updated this proposal to match the implementation in https://github.com/moby/buildkit/pull/499 . If you want to see the old proposal with different ARG names it is in https://gist.github.com/tonistiigi/79800156a5300336c24000cd78440a15

To enable configuring Dockerfile behavior based on the target platform or the platform of the node doing the build values of some predefined build args would be filled automatically.

These ars are:

```
TARGETPLATFORM
TARGETOS
TARGETARCH
TARGETVARIANT

BUILDPLATFORM
BUILDOS
BUILDARCH
BUILDVARIANT
```

`TARGETPLATFORM/BUILDPLATFORM` contain the normalized platform string value defined in https://github.com/containerd/containerd/blob/master/platforms/platforms.go#L17 . This is the same format accepted by `docker build --platform` or `FROM --platform=`.

`TARGET*` args are automatically filled with the value specified in `docker build --platform`. If no flag is specified this defaults to current platform. `FROM` command without `--platform` flag is equal to `FROM --platform=$TARGETPLATFORM` (this is the current behavior).

The `*OS/*ARCH/*VARIANT` vars are just different components of the main `*PLATFORM` value.

`BUILD*` args are automatically filled with the platform detected from current system. If system supports multiple platforms this is the default one.

The main use case for these variables would be to use them in multi-stage builds in combination with #31352. For example:

```
FROM alpine AS build-env-linux
...
FROM microsoft/nanoserver AS build-env-windows
...
FROM build-env-${TARGETOS}
```

Building this with `docker build -t myimage .` would now work both in linux and windows. Note that you would also need to use BuildKit for this to be efficient and skip unused commands.

They can also be used for cross compilation. In this example build stage is fixed to worker architecture but for the exported stage, alpine for target architecture is used.

```
FROM --from=$BUILDPLATFORM golang:alpine as build
ARG TARGETOS
ARG TARGETARCH
ENV GOOS=$TARGETOS GOARCH=$TARGETARCH
RUN go build ...

FROM alpine
COPY --from=build binary /bin
```

The build arguments are defined in global scope, meaning they can be used for argument substitution in `FROM` commands but do not automatically leak to your processes. To make them available inside the stage they need to be exposed with `ARG` like regular global scope arguments.

For more compatibility with older builders these args can be overriden by the user with `--build-arg` so if more control is needed or a new Dockerfile is build on old system it can still be used.

@dnephin @tiborvass

**Comment by duglin** [comment]
Not really a question about the issue itself, more on your example:
```
FROM alpine AS build-env-linux
...
FROM microsoft/nanoserver AS build-env-windows
...
FROM build-env-${DOCKERDOS}
```
that doesn't actually work does it? I mean w/o a mechanism to skip one of those two AS builds, one of them will probably fail since a single dockerd can't do both, right?

**Comment by dnephin** [comment]
I like this proposal. 

What would be the use case for `DOCKEROS` and `DOCKERARCH` ( the client variables) ?

Maybe `DOCKER_OS`, and `DOCKER_ARCH`, or `DOCKER_HOST_OS`, `DOCKER_HOST_ARCH` ?

**Comment by tonistiigi** [comment]
@duglin Yes, that example also needs the changes to move from linear execution to `DAG` and only execute the build stages that are actually required.

@dnephin For example there could be cases where your target is for building a binary that you wish to use/debug directly. Then `docker build --target docker-binary .` on osx could just build the OSX binary directly(same for windows) even if the daemon actually runs on Linux. For the underscores, I tried to use the same pattern that golang uses but do not have a strict preference.

**Comment by AkihiroSuda** [comment]
> Building this with docker build -t myimage . would now work both in linux and windows. This can be already achieved today but user would need to pass --buildarg DOCKEROS=windows manually.

Windows daemon is going to support LCOW.
So don't we still need to pass `DOCKERDOS` manually?

**Comment by tonistiigi** [comment]
@AkihiroSuda I guess there would still be one default for this variable. You could still always override it, even without `LCOW`. If the variable is not used in Dockerfile I guess the system would detect the method by the base image.

**Comment by tianon** [comment]
Won't this then need `DOCKERARM` (or some equivalent) too?  In a manifest list, we've got the full platform object, which has more expressivity than an individual image (`variant`, `features`, `os.version`, etc), but it seems like that amount of expressivity would actually be useful for a single image too (even if only to inform during the creation of a manifest list, or to decide whether the current system can run the image it's been asked to).

I think https://github.com/opencontainers/image-spec/issues/661 also has some tangental relevance (it's filed by the folks over at ARM trying to get some sanity to the way we tag/express ARM in manifests).

**Comment by tonistiigi** [comment]
I have updated this proposal to match implementation in moby/buildkit#499 

The change is to include arg with platform format defined in containerd and rename from `DOCKER/DOCKERD` to `TARGET/BUILD`

**Comment by andrewhsu** [comment]
@tonistiigi just needs a vndr to get this into moby?

**Comment by tonistiigi** [comment]
@andrewhsu In the BuildKit mode yes. But if we want this for legacy as well then implementation needs to be added. I think it makes sense in legacy as it is a sort of thing that can be added with almost full backward compatibility and we already support `FROM --platform` and `docker build --platform`.

**Comment by tonistiigi** [comment]
Implemented in moby/buildkit#499
